### PR TITLE
🎨 Add Typehints for `misc.py`

### PR DESCRIPTION
**Pull request by shaneahmed**
- Add Typehints for `misc.py`
- Includes commits from #619

### CODE REVIEW

File: tiatoolbox/utils/misc.py [hunk 1]
```diff
@@ -4,7 +4,7 @@
 import os
 import pathlib
 import zipfile
-from typing import IO, Dict, Optional, Tuple, Union
+from typing import IO, Dict, List, Optional, Tuple, Union
```
Comment (shaneahmed) [inline tiatoolbox/utils/misc.py:8]:
```suggestion
from os import PathLike
```

File: tiatoolbox/utils/misc.py [hunk 2]
```diff
@@ -45,11 +47,14 @@ def split_path_name_ext(full_path):
...
-def grab_files_from_dir(input_path, file_types=("*.jpg", "*.png", "*.tif")):
+def grab_files_from_dir(
+    input_path: PathLike,
+    file_types: Union[str, Tuple[str]] = ("*.jpg", "*.png", "*.tif"),
+) -> List[pathlib.Path]:
     """Grab file paths specified by file extensions.
 
     Args:
-        input_path (str or pathlib.Path):
+        input_path (PathLike):
```
Comment (shaneahmed) [inline tiatoolbox/utils/misc.py:57]:
```suggestion
        input_path (str | PathLike):
```

File: tiatoolbox/typing.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+import numpy as np
+
+Resolution = Union[Number, Tuple[Number, Number], np.ndarray]
+Units = Literal["mpp", "power", "baseline", "level"]
+NumPair = Tuple[Number, Number]
+IntPair = Tuple[int, int]
+Bounds = Tuple[Number, Number, Number, Number]
+IntBounds = Tuple[int, int, int, int]
+PathLike = Union[str, PathLike]
```
[the remark this answers is not in the document]
Comment (shaneahmed) [reply tiatoolbox/typing.py]: Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/620/commits/f729dc35407b7d347fd39bfb4a24dbe6856d4542

File: tiatoolbox/typing.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+from numbers import Number
+from os import PathLike
```
[the remark this answers is not in the document]
Comment (shaneahmed) [reply tiatoolbox/typing.py]: Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/620/commits/f729dc35407b7d347fd39bfb4a24dbe6856d4542

File: tiatoolbox/utils/misc.py (hunk 1 above)
Comment (shaneahmed) [reply tiatoolbox/utils/misc.py:9]: Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/620/commits/f729dc35407b7d347fd39bfb4a24dbe6856d4542

File: tiatoolbox/utils/misc.py (hunk 2 above)
Comment (shaneahmed) [reply tiatoolbox/utils/misc.py]: Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/620/commits/f729dc35407b7d347fd39bfb4a24dbe6856d4542

**Review by John-P — APPROVED** [review]

**Pull request merged**